databuilder/query_engines/query_model_convert_to_old.py

```diff
@@ -0,0 +1,224 @@
+    convert_node.cache_clear()
+    return old_cohort
+
+
+def convert_value(value):
+    if isinstance(value, new.Value):
+        return convert_value_node(value)
+    elif isinstance(value, new.Node):
+        return convert_node(value)
+    else:
+        assert False, f"Unhandled value {value!r}"
+
+
+@cache
+@singledispatch
+def convert_node(node):
+    assert False, f"Unhandled node type {type(node)}"
+
+
+def convert_value_node(node: new.Value):
+    value = node.value
+    if isinstance(value, frozenset) and any(isinstance(v, new.Code) for v in value):
+        return convert_codelist(value)
+    else:
+        return value
+
+
+@convert_node.register
+def convert_table(node: new.SelectTable):
+    return old.Table(node.name)
+
+
+@convert_node.register
+def convert_patient_table(node: new.SelectPatientTable):
+    return old.Row(old.Table(node.name), ("patient_id",), descending=False)
+
+
+@convert_node.register
+def convert_filter(node: new.Filter):
+    source = convert_node(node.source)
+    condition = node.condition
+
+    or_null = False
+    if isinstance(condition, new.Function.Or):
+        assert isinstance(condition.rhs, new.Function.IsNull)
+        assert condition.rhs.source == condition.lhs.lhs
+        condition = condition.lhs
+        or_null = True
+
+    if isinstance(condition, new.Function.IsNull):
+        condition = new.Function.EQ(condition.source, new.Value(None))
+
+    if isinstance(condition, new.Function.Not):
+        assert isinstance(condition.source, new.Function.In)
+        condition = condition.source
+        operator = "not_in"
+    else:
+        operator = OPERATOR_MAP[condition.__class__]
+
+    lhs, rhs = condition.lhs, condition.rhs
+    assert isinstance(lhs, new.SelectColumn)
+    assert lhs.source == node.source
+    column = column_name(lhs.name)
+
+    value = convert_value(rhs)
+
+    return old.FilteredTable(
+        source=source, column=column, operator=operator, value=value, or_null=or_null
+    )
+
+
+@convert_node.register
+def convert_select_column(node: new.SelectColumn):
+    source = convert_node(node.source)
+    column = column_name(node.name)
+    if isinstance(source, old.Row):
+        return old.ValueFromRow(source, column)
+    else:
+        return old.Column(source, column)
+
+
+@convert_node.register
+def convert_sort_and_select(node: new.PickOneRowPerPatient):
+    descending = node.position == new.Position.LAST
+    assert isinstance(node.source, new.Sort)
+    sort_columns = []
+    while isinstance(node.source, new.Sort):
+        assert isinstance(node.source.sort_by, new.SelectColumn)
+        sort_columns.insert(0, node.source.sort_by.name)
+        node = node.source
+    return old.Row(convert_node(node.source), tuple(sort_columns), descending)
+
+
+@convert_node.register
+def convert_combine_as_set(node: new.AggregateByPatient.CombineAsSet):
+    return convert_node(node.source)
```

Comment:
Is this right? Do we actually support this?

Reply:
We do support it in the existing query engine, just all in a very ad hoc and under-specified way. But if you try to filter using an IN condition and the thing on the RHS is an EventSeries (or `Column` in the old world) then we do the right thing. So `CombineAsSet` is a no-op in the conversion function — it's just functioning as a way to say: "Yes, I intend to do the thing I am about to do"

Reply:
Super.